src/flashops.c

```diff
@@ -72,11 +81,60 @@ int8_t writeDataflash(MSHW* ms, unsigned int translated_addr, uint8_t val)
 	return modified;
 }
 
+/* Generate and set a random serial number valid for Mailstation
+ *
+ * Disassembly of codeflash found that the verification process for serial num
+ * is that it starts at 0x7FFC8, is 16 bytes long, each character must be
+ * ASCII '0'-'9', 'A'-'Z', 'a'-'z', or '-'
+ * In all observed Mailstations, the last character is '-', but it does
+ * not seem to be enforced in Mailstation firmware. Nevertheless, the last char
+ * is set to a '-'
+ */
+void set_rnd_ms_ser(uint8_t *buf)
+{
+	int i;
+	uint8_t rnd;
+
+	srandom((unsigned int)time(NULL));
+	for (i = 0; i < 15; i++) {
+		do {
+			rnd = random();
+		} while (!isalnum(rnd));
+		*buf = rnd;
+		buf++;
+	}
+
+	*buf = '-';
+}
+
+/* Check if serial number is valid for MS
+ *
+ * Returns 1 if number is valid, 0 if invalid.
+ *
+ * Disassembly of codeflash found that the verification process for serial num
+ * is that it starts at 0x7FFC8, is 16 bytes long, each character must be
+ * ASCII '0'-'9', 'A'-'Z', 'a'-'z', or '-'
+ * In all observed Mailstations, the last character is '-', but it does
+ * not seem to be enforced in Mailstation firmware; it is not enforced here
+ */
+int check_ms_ser(uint8_t *buf)
```

Comment:
Same comment as above regarding the `buf` offset.

Reply:
Refactored, made this function accept ms_ctx and modify the dataflash from there. All self contained now.

And moved all DF related funcs to prefix with "df_"